### PR TITLE
Refactor name and id fields for attack graph

### PR DESCRIPTION
**Pull request by andrewbwm**
This also reworked some other existing fields.

### CODE REVIEW

File: tests/attackgraph/test_attackgraph.py [hunk 1]
```diff
@@ -274,7 +274,7 @@ def test_attackgraph_according_to_corelang(corelang_lang_graph, model):
 
 def test_attackgraph_regenerate_graph():
     """Make sure graph is regenerated"""
-    pass  # we don't have to test this atm tbh plz
```
Comment (mrkickling) [inline tests/attackgraph/test_attackgraph.py:277]: Sad to see this helpful comment go..

File: maltoolbox/attackgraph/attackgraph.py [hunk 2]
```diff
@@ -327,27 +359,26 @@ def attach_attackers(self):
...
                 reached_attack_steps = []
             )
-            self.attackers.append(ag_attacker)
+            self.add_attacker(attacker)
 
             for (asset, attack_steps) in attacker_info.entry_points:
                 for attack_step in attack_steps:
-                    attack_step_id = asset.name + ':' + attack_step
-                    ag_node = self.get_node_by_id(attack_step_id)
+                    full_name = asset.name + ':' + attack_step
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:371]: can we use attack_step.get_full_name here?

File: maltoolbox/attackgraph/attackgraph.py [hunk 3]
```diff
@@ -437,8 +467,10 @@ def _generate_graph(self):
                     step_expression)
 
                 for target in target_assets:
-                    target_node_id = target.name + ':' + attack_step
-                    target_node = self.get_node_by_id(target_node_id)
+                    target_node_full_name = target.name + ':' + attack_step
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:470]: can we use attack_step.get_full_name here?

File: maltoolbox/attackgraph/attackgraph.py [hunk 4]
```diff
@@ -164,10 +164,10 @@ class AttackGraph():
     """Graph representation of attack steps"""
     def __init__(self, lang_graph = None, model: Optional[Model] = None):
         self.nodes = []
-        self.id_to_node = {}  # to get nodes by id faster
-        self.full_name_to_node = {}  # to get nodes by full name faster
+        self._id_to_node = {}  # to get nodes by id faster
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:167]: Good with protected name, we should use it more (but not in this PR)

File: maltoolbox/attackgraph/attackgraph.py [hunk 5]
```diff
@@ -195,6 +195,7 @@ def _to_dict(self):
 
     def save_to_file(self, filename):
         """Save to json/yml depending on extension"""
+        logger.debug(f'Save attack graph to file\"{filename}\".')
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:198]: Might be worth to use the new format which is used in #49:
```
logger.debug('Save attack graph to file "%s".', filename) 
```

(I don't know how to make 'suggestions' in github)

File: maltoolbox/attackgraph/attackgraph.py [hunk 6]
```diff
@@ -297,14 +298,20 @@ def _from_dict(cls, serialized_object, model=None):
     @classmethod
     def load_from_file(cls, filename, model=None):
         """Create from json or yaml file depending on file extension"""
-        serialized_model = None
+        if model is not None:
+            logger.debug(f'Load attack graph from file\"{filename}\" with '
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:302]: same thing with logging format

File: maltoolbox/attackgraph/attackgraph.py [hunk 7]
```diff
@@ -5,7 +5,7 @@
 import logging
 import json
 
-from typing import Optional
+from typing import Optional, List
```
Comment (mrkickling) [inline maltoolbox/attackgraph/attackgraph.py:8]: Use lowercase list instead (following discussion in #44)

**Review by mrkickling — COMMENTED** [review]
lgtm, small remarks only

File: maltoolbox/attackgraph/attackgraph.py [hunk 8]
```diff
@@ -458,15 +497,34 @@ def regenerate_graph(self):
...
+        if logger.isEnabledFor(logging.DEBUG):
+            # Avoid running json.dumps when not in debug
+            logger.debug(f'Add node with id:{node_id}:\n' \
+                + json.dumps(node.to_dict(), indent = 2))
+
+        if node.id in self._id_to_node:
+            raise ValueError(f'Node index {node_id} already in use.')
+
+        node.id = node_id if node_id is not None else self.next_node_id
+        self.next_node_id = max(node.id + 1, self.next_node_id)
```
Comment (nkakouros) [inline maltoolbox/attackgraph/attackgraph.py:520]: This will leave gaps in the sequence of `id`s, but I guess that's ok?

**Review by nkakouros — COMMENTED** [review]
Logging statements use f-strings, they should be converted to `%` format strings.

**Review by mrkickling — APPROVED** [review]
lgtm

**Pull request merged**